File: Irrgarten/src/irrgarten/FuzzyPlayer.java
package irrgarten;
import java.util.ArrayList;

/**
 * Class for a fuzzy player. A fuzzy player is a player that makes decisions
 * based on a fuzzy logic system.
 */
public class FuzzyPlayer extends Player {
    
    /**
     * Constructor for the fuzzy player. Initializes the fuzzy player with the
     * number, intelligence, and strength of the player.
     * @param other Player to copy
     */
    public FuzzyPlayer(Player other){
        super(other.getNumber(), other.getIntelligence(), other.getStrength());
        this.setPos(other.getRow(), other.getCol());
    }
    
    /**
     * Moves the player in the given direction. The player chooses the direction
     * based on the fuzzy logic system.
     * @param direction The direction to move the player.
     * @param validMoves The valid moves for the player.
     * @return The direction the player moves.
     */
    @Override
    public Directions move(Directions direction, 
                            ArrayList<Directions> validMoves){
        Directions chosenDirection = this.move(direction, validMoves);
        Directions resultingDirection = 
            Dice.nextStep(chosenDirection, validMoves, this.getIntelligence());
        return resultingDirection;
    }
    
    /**
     * Attacks the opponent. The player chooses the with a dice and the sum of
     * weapons the player has.
     * @return The attack value of the player.
     */
    @Override
    public float attack(){
        return Dice.intensity(this.getStrength()) + this.sumWeapons();
    }
    
    /**
     * Defends the player. The player chooses the defense value with a dice and
     * the sum of shields the player has.
     * @return The defense value of the player.
     */
    @Override
    protected float defensiveEnergy(){
        return Dice.intensity(this.getIntelligence()) + this.sumShields();
    }
    
    /**
     * Returns the string representation of the fuzzy player.
     * @return The string representation of the fuzzy player.
     */
    @Override
    public String toString(){
        return "Fuzzy " + super.toString();
    }
    
    
}

File: Irrgarten_ruby/fuzzy_player.rb
# encoding: utf-8
require_relative 'player'
require_relative 'dice'
module Irrgarten
# Class that represents a player with a fuzzy intelligence. The player
# will have a preferred direction to move, but it will have a probability
# to move in a different direction.
# @see Player
class FuzzyPlayer < Player
    # Copy constructor
    def initialize(player)
        copy(player)
    end

    # The player will move in the preferred direction with a probability
    # based on its intelligence.
    # @return [Directions] the direction where the player will move
    def move(direction, valid_moves)
        preferred_direction = super(direction, valid_moves)
        Dice.next_step(preferred_direction, valid_moves, @intelligence)
    end

    # @return [Float] the defensive energy of the player
    def attack
        Dice.intensity(@strength) + self.sum_weapons
    end

    protected

    # @return [Float] the defensive energy of the player
    def defensive_energy
        Dice.intensity(@intelligence) + self.sum_shields
    end

    public

    # @return [String] a string representation of the player
    def to_s
        "Fuzzy #{super}"
    end

end
end
